### PR TITLE
fix: Modify validation to check Analysis args passed through RO spec

### LINKED ISSUE #1208: Analysis Template argument without default value can not be resolved in the rollout yaml
## Summary 

When AnalysisTemplate has an argument what does not have a default value or `valuefrom`. The controller's validation failed due to unresolved argument name.

```
# AnalysisTemplate which sleeps for a specified duration and exits with a specified exit-code
kind: AnalysisTemplate
apiVersion: argoproj.io/v1alpha1
metadata:
  name: sleep-job
spec:
  args:
  - name: duration
  - name: exit-code
    value: "0"
  metrics:
    - name: sleep-job
      count: 1
      provider:
        job:
          spec:
            template:
              spec:
                containers:
                  - name: sleep-job
                    image: nginx:1.19-alpine
                    command: [sh, -c, -x]
                    args: ["sleep {{args.duration}} && exit {{args.exit-code}}"]
                restartPolicy: Never
            backoffLimit: 0
---
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: rollout-bluegreen
spec:
  replicas: 3
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: rollout-bluegreen
  template:
    metadata:
      labels:
        app: rollout-bluegreen
    spec:
      containers:
        - name: rollouts-demo
          image: argoproj/rollouts-demo:blue
          imagePullPolicy: Always
          ports:
            - containerPort: 8080
              name: http
              protocol: TCP
  strategy:
    blueGreen:
      activeService: rollout-bluegreen-active
      previewService: rollout-bluegreen-preview
      autoPromotionEnabled: false
      prePromotionAnalysis:
        templates:
          - templateName: sleep-job
        args:
        - name: duration
          value: "5"
---
apiVersion: v1
kind: Service
metadata:
  name: rollout-bluegreen-active
spec:
  ports:
    - port: 80
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app: rollout-bluegreen
---
apiVersion: v1
kind: Service
metadata:
  name: rollout-bluegreen-preview
spec:
  ports:
    - port: 80
      targetPort: http
      protocol: TCP
      name: http
  selector:
    app: rollout-bluegreen

```

## Diagnostics

What version of Argo Rollouts are you running?

commit id: 912d3ac0097a5fc24932ceee532aa18bcc79944d

```
# Paste the logs from the rollout controller

# Logs for the entire controller:
kubectl logs -n argo-rollouts deployment/argo-rollouts

mespace=default rollout=rollout-bluegreen
INFO[2021-05-20T16:15:19-04:00] Reconciliation completed                      generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen time_ms=1.261676
INFO[2021-05-20T16:15:39-04:00] Started syncing rollout                       generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen
ERRO[2021-05-20T16:15:39-04:00] The Rollout "rollout-bluegreen" is invalid: spec.strategy.blueGreen.prePromotionAnalysis.templates: Invalid value: "templateNames: [sleep-job]": args.duration was not resolved  namespace=default rollout=rollout-bluegreen
INFO[2021-05-20T16:15:39-04:00] Reconciliation completed                      generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen time_ms=1.2937
INFO[2021-05-20T16:15:59-04:00] Started syncing rollout                       generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen
ERRO[2021-05-20T16:15:59-04:00] The Rollout "rollout-bluegreen" is invalid: spec.strategy.blueGreen.prePromotionAnalysis.templates: Invalid value: "templateNames: [sleep-job]": args.duration was not resolved  namespace=default rollout=rollout-bluegreen
INFO[2021-05-20T16:15:59-04:00] Reconciliation completed                      generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen time_ms=1.0176180000000001
INFO[2021-05-20T16:16:19-04:00] Started syncing rollout                       generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen
ERRO[2021-05-20T16:16:19-04:00] The Rollout "rollout-bluegreen" is invalid: spec.strategy.blueGreen.prePromotionAnalysis.templates: Invalid value: "templateNames: [sleep-job]": args.duration was not resolved  namespace=default rollout=rollout-bluegreen
INFO[2021-05-20T16:16:19-04:00] Reconciliation completed                      generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen time_ms=1.3263120000000002
INFO[2021-05-20T16:16:39-04:00] Started syncing rollout                       generation=2 namespace=default resourceVersion=33424 rollout=rollout-bluegreen
ERRO[2021-05-20T16:16:39-04:00] The Rollout "rollout-bluegreen" is invalid: spec.strategy.blueGreen.prePromotionAnalysis.templates: Invalid value: "templateNames: [sleep-job]": args.duration was not resolved  namespace=default rollout=rollout-bluegreen



# Logs for a specific rollout:
kubectl logs -n argo-rollouts deployment/argo-rollouts | grep rollout=<ROLLOUTNAME>
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by khhirani**
Resolves #1208 

Signed-off-by: khhirani <kareena.hirani@gmail.com>

### CODE REVIEW

**Comment by huikang** [comment]
Thanks for fixing.
Can we add some unit test or e2e test to avoid regression?

File: pkg/apis/rollouts/validation/validation_references.go [hunk 1]
```diff
@@ -103,14 +107,9 @@ func ValidateAnalysisTemplatesWithType(rollout *v1alpha1.Rollout, templates Anal
...
 	}
 
-	flattenedTemplate, err := analysisutil.FlattenTemplates(templates.AnalysisTemplates, templates.ClusterAnalysisTemplates)
 	templateNames := GetAnalysisTemplateNames(templates)
 	value := fmt.Sprintf("templateNames: %s", templateNames)
-	if err != nil {
-		allErrs = append(allErrs, field.Invalid(fldPath, value, err.Error()))
-		return allErrs
-	}
-	err = analysisutil.ResolveArgs(flattenedTemplate.Spec.Args)
```
Comment (huikang) [inline pkg/apis/rollouts/validation/validation_references.go:113]: it seems this method can be removed as this is the only place to use it.

**Review by jessesuen — CHANGES_REQUESTED** [review]
Tests?

**Comment by khhirani** [comment]
@jessesuen I modified the existing tests to cover the change

File: rollout/controller.go [hunk 2]
```diff
@@ -636,6 +636,7 @@ func (c *rolloutContext) getReferencedRolloutAnalyses() (*[]validation.AnalysisT
 			if err != nil {
 				return nil, err
 			}
+			templates.Args = blueGreen.PrePromotionAnalysis.Args
```
Comment (jessesuen) [inline rollout/controller.go:639]: 1. Are you modifying the AnalysisTemplate which was pulled off the informer store (lister)? Objects received form the store should not be modified. See:

https://github.com/kubernetes/sample-controller/blob/b8d9e8c247129e53962d0dcfc08a4e8b47477318/controller.go#L322-L324

2. It seems we are clobbering the entire arguments of the template with the Arguments of the Rollout. Does this change support default values?

Comment (khhirani) [reply rollout/controller.go:639]: > It seems we are clobbering the entire arguments of the template with the Arguments of the Rollout.

Can you elaborate on this? I thought we did this anyway when we created the AnalysisRun from the templates (i.e. combine the templates’ args with the AnalysisRunArguments passed through the RO spec)

Comment (khhirani) [reply rollout/controller.go:639]: > Are you modifying the AnalysisTemplate which was pulled off the informer store (lister)?

I don't believe so. I'm copying the AnalysisRunArguments from the Rollout Spec into a `AnalysisTemplatesWithType` object, which is a struct I created for validation. I pass the templates and the args from the RO spec into `NewAnalysisRunFromTemplates` for validation, which doesn't seem to modify the templates.

Comment (jessesuen) [reply rollout/controller.go:639]: > Can you elaborate on this? 

Can you confirm that this is the behavior?

AnalysisTemplate
```yaml
args:
- name: aaa
  value: foo     # default value
- name: bbb  # mandatory supplied value
- name: ccc
  value: baz
```

Args supplied by rollout (note that aaa is omitted)
```yaml
args:
- name: bbb
  value: 11111
- name: ccc
  value: 2222
```

The AnalysisRun should be created with

```yaml
args:
- name: aaa
  value: foo
- name: bbb
  value: 11111
- name: ccc
  value: 2222
```

Comment (jessesuen) [reply rollout/controller.go:639]: > I don't believe so. I'm copying the AnalysisRunArguments from the Rollout Spec into a AnalysisTemplatesWithType object, which is a struct I created for validation.

Yes, but AnalysisTemplatesWithType contains a list of **_pointers to AnalysisTemplates and ClusterAnalysisTemplates_** which were pulled off the store, which you are directly modifying.

Comment (jessesuen) [reply rollout/controller.go:639]: > I don't believe so. I'm copying the AnalysisRunArguments from the Rollout Spec into a AnalysisTemplatesWithType object, which is a struct I created for validation

I took a closer look. You're right. I mistakenly thought it was returning AnalysisTemplates

Comment (khhirani) [reply rollout/controller.go:639]: @jessesuen Can I merge this in then?

Comment (jessesuen) [reply rollout/controller.go:639]: My first concern is not valid, but can you answer my second concern?

Comment (khhirani) [reply rollout/controller.go:639]: @jessesuen Yes, I confirm that the behavior in [this comment](https://github.com/argoproj/argo-rollouts/pull/1215#discussion_r639044608) is what happens.

**Review by jessesuen — APPROVED** [review]

**Pull request merged**